### PR TITLE
Introduce `ExtensionsManagerProtocol` for type hint of extension/trigger argument

### PR DESCRIPTION
**Pull request by asi1024**

### CODE REVIEW

**Comment by kmaehashi** [comment]
Thanks! Could you use `typing.Protocol`?

**Comment by asi1024** [comment]
`typing.Protocol` is supported in Python 3.8 or later.

**Comment by asi1024** [comment]
@kmaehashi Fixed to use `typing_extensions.Protocol`. PTAL.

File: setup.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
                 'development in PyTorch.',
     author='Preferred Networks, Inc.',
     license='MIT License',
-    install_requires=['numpy', 'torch'],
+    install_requires=['numpy', 'torch', 'typing-extensions'],
```
Comment (kmaehashi) [inline setup.py:17]: Better to add a version restriction (`>=3.10`)?

Comment (asi1024) [reply setup.py]: Fixed.

File: pytorch_pfn_extras/training/manager.py [hunk 2]
```diff
@@ -57,6 +58,45 @@ def epoch_detail(self) -> float:
     def _iters_per_epoch(self) -> int:
         return self._manager._iters_per_epoch
 
+    @property
+    def models(self) -> Dict[str, torch.nn.Module]:
+        return self._manager.models
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:63]: As for execution trigger, this must not be available.

File: pytorch_pfn_extras/training/manager.py [hunk 3]
```diff
@@ -57,6 +58,45 @@ def epoch_detail(self) -> float:
...
+
+    @property
+    def raw_models(self) -> Mapping[str, torch.nn.Module]:
+        return self._manager.raw_models
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:67]: ditto

File: pytorch_pfn_extras/training/manager.py [hunk 4]
```diff
@@ -57,6 +58,45 @@ def epoch_detail(self) -> float:
...
+
+    @property
+    def optimizers(self) -> Mapping[str, torch.optim.Optimizer]:
+        return self._manager.optimizers
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:71]: ditto

**Review by kmaehashi — CHANGES_REQUESTED** [review]

**Comment by kmaehashi** [comment]
Sorry my wording was ambiguous, I expected these APIs (`models` etc) to raise an error for the proxy.

**Comment by asi1024** [comment]
PTAL!

File: pytorch_pfn_extras/training/_manager_protocol.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    def epoch(self) -> int:
+        ...
+
+    @property
+    def epoch_detail(self) -> float:
+        ...
+
+    @property
+    def _iters_per_epoch(self) -> int:
+        ...
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/_manager_protocol.py:28]: Let's remove this from public API!

**Review by kmaehashi — CHANGES_REQUESTED** [review]
Thanks, LGTM except minor comments!

File: pytorch_pfn_extras/training/_manager_protocol.py (hunk 5 above)
Comment (kmaehashi) [inline pytorch_pfn_extras/training/_manager_protocol.py:28]: As discussed remove this from public API.

**Review by kmaehashi — APPROVED** [review]
LGTM!

**Pull request merged**